### ISSUE TITLE
Pagination with cache is still goofy

### ISSUE THREAD

**Issue by wesbos** [wesbos/advanced-react-rerecord#55]
getting help from the big guns at apollo for this one

**Comment by wesbos** [comment]
Still waiting on the folks at Apollo, but they are working on a fix or a solution for us.

Just for reference the issue is this:

We have 4 items per page.

When I visit a page, those 4 items are slotted into the cache.

When you add or remove an item, that cache needs to be updated. 

Now we can control the cache in two ways:

1. Each query has it's own cache:

allItems(limit:4, offset: 4)

2. All items in a single cache, this removes the `limit` and `cache` variables from the query and we manage the field ourselves with a typePolicy. 

Both have an issue. Let's say we add an item:

1. With method 1, all the items need to be evicted or re-fetched
  1. There is currently no way to evict all items of a type. You can either delete 1 item from the cache, or blow the entire apollo cache away.
  2. With refetching queries, I'd need to supply an array of queries that need to be re-fetvched. But when an item is added, we don't know how many queries are in the cache, how many pages there are, what the offsets are etc.... And we could possibly have hundreds of pages in the cache

2. With method 2, we can put all the items in one cache, so this means when we delete or add an item, we simply just add or remove it from the cache in the slot it needs. Great! This actually works well, _BUT_ this means we have to use a type policy for `allItems`, and then we aren't able to use the `allItems` query for anything else on the page. In our case we need it for search. So if I search for "shoes", the cache is now filled with items that match the search query and all our pagination data is gone.

So that's where we are at right now... I'm very open to having approached this absolutely wrong because I can't believe I'm the only one with this issue, but from talking to @kkemple it seems we are doing it right! 

So that is where we are at! Huge thanks for the apollo folks for helping with this. 

Once this is finished we are GOOD TO GOOOOOO

**Comment by RangerCoder99** [comment]
Hope there is a process on this issue, Also did you pick one backend to use yet? Like you still seem to be going between keystone and Strapi.

**Comment by Asjas** [comment]
Hey @RangerCoder99 

Wes mentioned on Twitter that he chose Keystone 🙂

**Comment by gmwill934** [comment]
How about using RQ?

**Comment by Asjas** [comment]
Wes is using Apollo Client 3 in this course 😃  @gmwill934

**Comment by wesbos** [comment]
thanks @Asjas! 

Yes I'm just going ahead with the recording. The folks at apollo have been super helpful but I can't wait any more. There might be bugs in the pagination which can be remedied with a video update if they ever find a solution. 

Recording is happening right now!

**Comment by wesbos** [comment]
Oh and Keystone vs Strapi: 

they are very very close, Strapi has a better UI IMO, keystone has a bit better access control. also keystone is being re-written in Next.js (will be included in the course), so customizing the backend will be super easy.

**Comment by adamtang79** [comment]
Hi @wesbos , i'm tired of rest and trying graphql after some recommendation. Initially I was from ws then grpc. I've got the feeling that graphql are just another type of nosql. Can I do short call as well as long call? eg. I had used loopback rest to call a function to process things up to 6 hours. I think method 1 to blow the whole cache is acceptable, of course, smart invalidation will be nice to have. When are your new recordings released in your planning?

**Comment by esnakker** [comment]
> Oh and Keystone vs Strapi:
> 
> they are very very close, Strapi has a better UI IMO, keystone has a bit better access control. also keystone is being re-written in Next.js (will be included in the course), so customizing the backend will be super easy.

Have been playing around with Strapi since the beta and I must say: I really love it! It's super flexible and the access control is also very granular (per role and for any single content type) you have a media library that can be linked with an S3 bucket and I also really like the Component feature (formerly dropzones). They enable you to build flexible layout from the backend (like gutenberg in WP).

Would be absolutely great to watch a course were you are using it for the backend. Strapi is also very high in demand by many (ongoing) webdevelopers. I thing it will be a big think in 2021.
